build: enable goimports and varcheck linters

thomasmodeneis opened the pull request:
This is WIP, the list of linting issues is quite big as follows:
I will start working on fixing the issues now.

Complete lint report: https://gist.github.com/thomasmodeneis/41e166b9f2fed52ccf29ad123f03a18a

I see that the actual linting is done inside the `ci.go`, why is that ? 

What are the benefits of coupling a linting task inside a go file ? and not doing it on the makefile with simple two liner command ?

Thanks.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

fjl commented:
We use build/ci.go because it is platform-independent (the Makefile doesn't work on Windows) and because we can share setup code for GOPATH, etc across built tasks. 

It looks like you are trying to enable the "varcheck" linter. Please do that by adding to the list in buidl/ci.go.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

 * Commit: b2e72fccd6ee1b787f33454a8fde4382b833be06
  * Commit subjects should be kept under 100 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

thomasmodeneis commented:
@fjl I've updated buidl/ci.go and fixed all the lint errors.
I also added little script for the gofmt as I could not find anything on the project already.

There was some issues with goimports that I had to revert, so there is a large list of changed files that in the end the change was just a minimal gofmt.

Please let me know if there is anything else that need to be done.

fjl commented:
This PR is well-intentioned and I like that you are taking on this effort. There are a few things that should be changed though:

The splitting of imports the way you did should be avoided. We use the default formatting introduced by goimports where you have two 'import groups': standard library and everything else. You split imports into three groups: standard library, external code and project code. We won't be able to stick to this convention because that's not what goimports does by default. Please revert the changes to import lists.

Please note that there is a dedicated Travis build job for running lint. We don't need to add it to the go 1.9.x linux builder.

Finally, when fixing the errors reported by varcheck, you can just delete the variables instead of commenting them out.

Thank you.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

 * Commit: b2e72fccd6ee1b787f33454a8fde4382b833be06
  * Commit subjects should be kept under 100 characters

 * Commit: f77e0b46e0e3bd192a2d89aa011b59e032c74bc5
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

thomasmodeneis commented:
> This PR is well-intentioned and I like that you are taking on this effort
Thanks, I'm glad you like it.

> Please revert the changes to import lists.
Hi @fjl Sorry for splitting the imports that way. This is the way I do in other projects. I added `build/goimports.sh` that runs gofmt and goimports on standard way, so this won't happen again as long as this script is used.

> We don't need to add it to the go 1.9.x linux builder.
I also removed the lint from the go build, sorry I missed the dedicated task for lint on travis.

> you can just delete the variables instead of commenting them out.
I did not delete the vars because I did not know this is what you wanted, so I did it now.


Please let me know if there is anything else that need to be done.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

 * Commit: b2e72fccd6ee1b787f33454a8fde4382b833be06
  * Commit subjects should be kept under 100 characters

 * Commit: f77e0b46e0e3bd192a2d89aa011b59e032c74bc5
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

fjl commented:
This looks a lot better now. Thank you so much!

I just noticed that you have also added the vendorcheck linter. We don't really need this linter because missing dependencies will cause a build failure on CI (it doesn't run go get). Please remove vendorcheck. We can merge then.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

 * Commit: b2e72fccd6ee1b787f33454a8fde4382b833be06
  * Commit subjects should be kept under 100 characters

 * Commit: f77e0b46e0e3bd192a2d89aa011b59e032c74bc5
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

thomasmodeneis commented:
> This looks a lot better now. Thank you so much!

I'm happy to be able to help out.

> Please remove vendorcheck. 

Sure, vendorcheck is removed as you requested.

karalabe commented on p2p/discv5/net.go:
```diff
@@ -828,11 +827,10 @@ type nodeEvent uint
 	// Packet type events.
 	// These correspond to packet types in the UDP protocol.
-	pingPacket
+	pingPacket = iota
```
This needs to start from 1.

karalabe commented on p2p/peer.go:
```diff
@@ -47,8 +47,6 @@ const (
 	pingMsg      = 0x02
 	pongMsg      = 0x03
-	getPeersMsg  = 0x04
-	peersMsg     = 0x05
```
Why are these unused? Are we sure this is correct here @fjl ?

karalabe commented on core/tx_pool.go:
```diff
@@ -39,7 +39,6 @@ const (
 	// chainHeadChanSize is the size of channel listening to ChainHeadEvent.
 	chainHeadChanSize = 10
 	// rmTxChanSize is the size of channel listening to RemovedTransactionEvent.
-	rmTxChanSize = 10
```
Please drop comment too.

karalabe commented on build/goimports.sh:
```diff
@@ -0,0 +1,18 @@
+GOFMT="gofmt -s -w";
+GOIMPORTS="goimports -w";
+find_files | xargs $GOFMT;
+find_files | xargs $GOIMPORTS;
```
What's the purpose of this file?

thomasmodeneis replied:
One of the reasons of linting errors was due to having code not properly formatted and not properly imported. There was no standard set for that. Now there is.

karalabe replied:
`goimports` is pretty much the standard tool used by any Go development environment. I don't really see the value of providing an extra utility method in our repo that is covered by the developer's editor anyway. I completely agree that the linter should enforce it, but its the developer's job to configure their dev environment, not ours.

thomasmodeneis replied:
When I started to work on this issue, I found that the devs are not really doing what you said is their job... 
```
consensus/consensus.go:1::warning: file is not goimported (goimports)
swarm/fuse/fuse_dir.go:1::warning: file is not goimported (goimports)
swarm/fuse/swarmfs.go:1::warning: file is not goimported (goimports)
```

So I ran the goimports the way "I think" is the best way, splitting the imports in 3. Just to find out that this is not the proffered way here. 
Also, running `goimports ./..` throws some errors, that I'm not sure how to fix, or even if they are really errors, so I added the script:
```
goimports -w ./..
../go-ethereum/crypto/bn256/bn256_fast.go:26:9: expected type, found '='
../go-ethereum/crypto/bn256/bn256_fast.go:30:9: expected type, found '='
../go-ethereum/crypto/bn256/bn256_fast.go:34:2: expected declaration, found 'return'
../go-ethereum/crypto/bn256/bn256_slow.go:26:9: expected type, found '='
../go-ethereum/crypto/bn256/bn256_slow.go:30:9: expected type, found '='
../go-ethereum/crypto/bn256/bn256_slow.go:34:2: expected declaration, found 'return'
```
So I added a "default" script to get this sorted and leave no room for errors.

karalabe commented on build/ci.go:
```diff
@@ -329,18 +329,22 @@ func doLint(cmdline []string) {
 		"--enable=gofmt",
 		"--enable=misspell",
 		"--enable=goconst",
+		"--deadline=5m",
```
Please don't add a deadline here. Only fast linters should be included in this batch that finish in a couple seconds. For slower linters, there's a second pass below.

thomasmodeneis replied:
Travis build fails randomly and report false alerts. So I increased the deadline to avoid having a false positive.

karalabe replied:
Please don't. The point of these linter steps is to run (and fail) fast. Any linter step requiring more than a few secs should be placed below, in the slow set.

thomasmodeneis replied:
Is Travis that is slow. Is a shared env. 
But ok, as you wish.

karalabe commented on Makefile:
```diff
@@ -37,6 +37,9 @@ ios:
 	build/env.sh go run build/ci.go test
 
+lint: ## Run linters. Use make install-linters first.
+	build/env.sh go run build/ci.go lint
```
There's no `install-linters` step.

fjl replied:
Oh wait, it's not removed? It didn't show up for me until now.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

 * Commit: b2e72fccd6ee1b787f33454a8fde4382b833be06
  * Commit subjects should be kept under 100 characters

 * Commit: f77e0b46e0e3bd192a2d89aa011b59e032c74bc5
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

 * Commit: b2e72fccd6ee1b787f33454a8fde4382b833be06
  * Commit subjects should be kept under 100 characters

 * Commit: f77e0b46e0e3bd192a2d89aa011b59e032c74bc5
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

fjl commented on p2p/discv5/net.go:
```diff
@@ -828,11 +827,10 @@ type nodeEvent uint
 //go:generate stringer -type=nodeEvent
 
 const (
-	invalidEvent nodeEvent = iota // zero is reserved
```
I just noticed another issue here: the nodeEvent String method is created by stringer. Please re-run `go generate ./p2p/discv5` to update nodeevent_string.go.

thomasmodeneis replied:
I'm running into a error when running stringer:
```go generate -v ./p2p/discv5
p2p/discv5/database.go
p2p/discv5/database_test.go
p2p/discv5/net.go
stringer: checking package: database.go:23:2: could not import bytes (reading export data: /usr/local/go/pkg/linux_amd64/bytes.a: invalid encoding format in export data: got 'v'; want 'c' or 'd')
p2p/discv5/net.go:827: running "stringer": exit status 1
```
I tried to find a way out, but so far I found nothing, any ideas ?

thomasmodeneis replied:
@fjl sorry for the delay, I got caught in other issue to work on. Just re-generated the nodeevent with stringer as you requested.
Please let me know if anything else is missing.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ebf6df049d55f15477b86ac6178c5607c8504139
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8889632abe74560afb177cdb376a0237182e50cb
  * Commits must be prefixed with the package(s) they modify

 * Commit: c41a7cc5fad942091c52565785d9a85eb57912a7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9671bfcaa0caccbb9b66457538dcd8bdcac2274a
  * Commits must be prefixed with the package(s) they modify

 * Commit: a48f2a45b9bcf936c1e47ce61b9489686b31ac84
  * Commits must be prefixed with the package(s) they modify

 * Commit: dea717b9f4c5ce1a977da20cd949cda9b7bf13d3
  * Commits must be prefixed with the package(s) they modify

 * Commit: b2e72fccd6ee1b787f33454a8fde4382b833be06
  * Commit subjects should be kept under 100 characters

 * Commit: f77e0b46e0e3bd192a2d89aa011b59e032c74bc5
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com
